### DISCUSSION TITLE
Massive `PRNGKey` allocations

### DISCUSSION THREAD

**Discussion by femtomc** [jax-ml/jax#12601] [Q&A] [answered]
I'm running a large computation with scans, with `vmap` inside of scans, etc.

Before the main `scan` computation, I pre-allocate a set of `sub_keys`. However, I'm still running into OOM errors (massive allocations) for `jax.random.split` calls in the middle of my computation. The below is one of 8 (!) buffers. So massive OOM.

<img width="712" alt="image" src="https://user-images.githubusercontent.com/34410036/193326896-72257596-e82a-483a-9c58-fba6b8fa3f14.png">

How do I avoid this? I'm doing lots of sampling, so it seems unavoidable to call `jax.random.split` -- on the other hand, I wish there was some way to evolve a set of keys in `scan` without newly allocating.

Also, what are best practices for keeping tracking of keys and reducing key allocations?

**Comment by froystig** [comment, marked as answer]
Can you make use of [`jax.random.fold_in`](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.fold_in.html) within the scan to avoid upfront allocation via split? For example, if you increment an integer counter between each scan iteration, you can fold that in to a fixed "root" scan key to produce a local subkey for each scan iteration.

**Reply by femtomc** [reply]
Great tip! I'll try immediately.
